### PR TITLE
Create sub directories for ThirdPartyAudit dependency metadata

### LINKED ISSUE #16840: [BUG] thirdPartyAudit task fails when copying duplicate files
### Describe the bug

Plugin dependencies which use identical file names for their metadata will cause the `thirdPartyAudit` task to fail for the plugin with the following error:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':plugins:transport-grpc:thirdPartyAudit'.
> Could not copy zip entry '/Users/carrofin/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/33.2.1-jre/818e780da2c66c63bbb6480fef1f3855eeafa3e4/guava-33.2.1-jre.jar!META-INF/LICENSE' to '/Users/carrofin/fdev/tmp/OpenSearch/plugins/transport-grpc/build/precommit/thirdPartyAudit/thirdPartyAudit/META-INF/LICENSE'.
   > /Users/carrofin/fdev/tmp/OpenSearch/plugins/transport-grpc/build/precommit/thirdPartyAudit/thirdPartyAudit/META-INF/LICENSE (Is a directory)

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Get more help at https://help.gradle.org.
```

In this case dependency 1 attempts to write a LICENSE file to `plugins/<my-plugin>/build/precommit/thirdPartyAudit/thirdPartyAudit/META-INF` while dependency 2 populates a LICENSE directory at the same location.

### Related component

Build

### To Reproduce

Note: It appears this specific minimal example does not reproduce on case sensitive file systems. At some point gradle normalizes the capitalization of `LICENSE` the file which in actuality is lowercase when written to disk.

Branch with minimal example: https://github.com/finnegancarroll/OpenSearch/tree/dbg-plugin-3prtyaud

Reproduce with:
```
./gradlew precommit -x :plugins:transport-grpc:testingConventions
```

### Expected behavior

`thirdPartyAudit` task should handle this collision gracefully and pass.
Proposed solution - Create sub directories for each dependency: #16844

### PR DESCRIPTION
**Pull request by finnegancarroll**
### Description
The `thirdPartyAudit` task writes all dependency metadata to the same base directory of `build/precommit/thirdPartyAudit/thirdPartyAudit/META-INF`. When two separate dependencies attempt to write a file and directory at the same location the task fails with an error.

See #16840.

This PR handles collisions by appending a counter to duplicates.
i.e. dup-1.txt, dup-2.txt, ...

### Related Issues
Resolves #16840

### Check List
- [ ] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
> When two separate dependencies attempt to write a file and directory at the same location the task fails with an error.

The task should be run per module, each having own `build` directory, with no overlaps (I think). Does some modules share the build folder?

**Comment by reta** [comment]
> In the minimal example linked in the issue these two randomly chosen dependencies happen to have jars which unzip a `META-INF/LICENSE` directory and `META-INF/LICENSE` file respectively.

Oh I see, thank you, `guava` as a dependency is causing troubles everywhere

File: buildSrc/src/main/java/org/opensearch/gradle/precommit/ThirdPartyAuditTask.java [hunk 1]
```diff
@@ -309,6 +309,25 @@ private void extractJars(Set<File> jars) {
         jars.forEach(jar -> {
             FileTree jarFiles = getProject().zipTree(jar);
             getProject().copy(spec -> {
+                spec.eachFile(details -> {
```
Comment (reta) [inline buildSrc/src/main/java/org/opensearch/gradle/precommit/ThirdPartyAuditTask.java:312]: Do you think the cleanest approach would be to extract each dependency into own sub-folder all the time so we won't be having the names clash in the first place?

Comment (finnegancarroll) [reply buildSrc/src/main/java/org/opensearch/gradle/precommit/ThirdPartyAuditTask.java]: Moved to creating sub directories.

File: buildSrc/src/main/java/org/opensearch/gradle/precommit/ThirdPartyAuditTask.java [hunk 2]
```diff
@@ -301,16 +300,26 @@ private void logForbiddenAPIsOutput(String forbiddenApisOutput) {
...
+     * @return File set of extracted jars
+     */
+    private Set<File> extractJars(Set<File> jars) {
+        Set<File> extractedJars = new TreeSet<>();
         File jarExpandDir = getJarExpandDir();
         // We need to clean up to make sure old dependencies don't linger
         getProject().delete(jarExpandDir);
 
         jars.forEach(jar -> {
+            String jarPrefix = jar.getName().replaceAll("-\\d+\\.\\d+\\.\\d+\\.jar$", "");
```
Comment (reta) [inline buildSrc/src/main/java/org/opensearch/gradle/precommit/ThirdPartyAuditTask.java:316]: Could we only strip `jar`? There are many artifacts that do not follow this scheme, notably Lucene snapshots : `lucene-core-10.2.0-SNAPSHOT-e732f3e.jar`, ...

File: buildSrc/src/main/java/org/opensearch/gradle/precommit/ThirdPartyAuditTask.java (hunk 1 above)
Comment (reta) [reply buildSrc/src/main/java/org/opensearch/gradle/precommit/ThirdPartyAuditTask.java]: Thank you

**Comment by finnegancarroll** [comment]
[org.opensearch.index.engine.NRTReplicationEngineTests.testGetSegmentInfosSnapshotPreservesFilesUntilRelease is flaky](https://github.com/opensearch-project/OpenSearch/issues/15817)

**Review by reta — APPROVED** [review]

**Pull request merged**